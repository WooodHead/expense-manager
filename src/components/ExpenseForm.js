import React from 'react';
import moment from 'moment';
import 'react-dates/initialize';
import { SingleDatePicker } from 'react-dates';

const now = moment();
//console.log(now.format());

export default class ExpenseForm extends React.Component{
    constructor(props){
        super(props);

        this.state = {
            id: props.expense ? props.expense.id : '',
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? props.expense.amount.toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: ''
        };
        //console.log('Id is'+ this.state.id);
    }
    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description }));
    }
    onNoteChange = (e) => {
        const note = e.target.value; // or we can use e.persist(); and use the e.target while setting state
        this.setState(() => ({ note }));
    }
    onAmountChange = (e) => {
        const amount = e.target.value;

        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){
           this.setState(() => ({ amount }));
        }
    }
    onDateChange = (createdAt) => {
        if(createdAt){
            this.setState(() => ({ createdAt }));
        }
    }
    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }));
    }
    onSubmit = (e) => {
        e.preventDefault();

        if(!this.state.description || !this.state.amount){
            // Set an error message
            const error = this.state.error;
            this.setState(() => ({ error: 'Please enter a valid description and amount' }));
        }
        else{
            // Clear the error
            this.setState(() => ({ error: '' }));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount,10),
                note: this.state.note,
                createdAt: this.state.createdAt.valueOf()
            })
        }

    }
    render(){
        return(
            <div>
                {this.state.error !== null && <p>{this.state.error}</p>}
                <h4> Expense Form </h4>
                <form onSubmit={this.onSubmit}>
                    <input 
                        type="text" 
                        placeholder="Description" 
                        autoFocus 
                        value={this.state.description} 
                        onChange={this.onDescriptionChange}
                    />
                    <input 
                        type="text" 
                        placeholder="Amount" 
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />
                    <SingleDatePicker 
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    />
                    <textarea 
                        placeholder="Add a note for the expense"
                        value={this.state.note}
                        onChange={this.onNoteChange}
                    >
                    </textarea>
                    <button>
                    {this.state.id !== "" ? 'Update Expense' : 'Add Expense'}
                    </button>
                </form>
            </div>
        )
    }
}