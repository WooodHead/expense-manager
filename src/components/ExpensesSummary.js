import React from 'react';
import {connect} from 'react-redux';
import numeral from 'numeral';
import selectExpenses from '../selectors/expenses';
import expensesTotal from '../selectors/expenses-total';

export const ExpensesSummary = ({ expensesCount, expensesTotal}) => {
    const expenseWord = expensesCount === 1 ? 'expense' : 'expenses';
    return (
        <div>
            <h1> Viewing {expensesCount} {expenseWord} totalling {numeral(expensesTotal).format('$0,00.00')}.</h1>
        </div>
    );
};

const mapStateToProps = (state) => {
    const visibleExpenses = selectExpenses(state.expenses, state.filters);

    return {
        expensesCount: visibleExpenses.length,
        expensesTotal: expensesTotal(visibleExpenses)
    }
};

export default connect(mapStateToProps)(ExpensesSummary);