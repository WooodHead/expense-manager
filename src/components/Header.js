import React from 'react';
import {NavLink} from 'react-router-dom';
import { startLogout } from '../actions/auth';
import {connect} from 'react-redux';


export const Header = ({ startLogout }) =>(
    <header>
        <h3>Expense Manager</h3>
        <NavLink to="/dashboard" activeClassName='isActive' exact={true}>Dashboard</NavLink>
        <NavLink to="/create" activeClassName='isActive' exact={true}>Create Expense</NavLink>
        <button onClick={startLogout}>Logout</button>
    </header>
);

const mapDipatchToProps = (dispatch) => ({
    startLogout: () => dispatch(startLogout())
});

export default connect(undefined, mapDipatchToProps)(Header);