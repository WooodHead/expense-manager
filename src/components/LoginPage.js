import React from 'react';
import { connect } from 'react-redux';
import { startLogin } from '../actions/auth';

export const LoginPage = ({ startLogin }) => (
    <div className="box-layout">
        <div className="box-layout__box">
            <h1 className="box-layout__title">Expense Manager</h1>
            <p>Manage your money the smart way!!</p>
            <button className="button" onClick={startLogin}>Login with Google</button>
        </div>
    </div>
);

const mapDipatchToProps = (dispatch) => ({
    startLogin: () => dispatch(startLogin())
});

export default connect(undefined, mapDipatchToProps)(LoginPage);