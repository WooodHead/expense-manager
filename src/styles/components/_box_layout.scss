.box-layout {
    background: url('/images/banner.jpeg');
    height: 100vh;
    width: 100vw;
    display: flex;
    justify-content: center;
    align-items: center;
    background-size: cover;
}
.box-layout__box{
    background: fade-out($color: #fff, $amount: 0.15);
    border-radius: 3px;
    padding: $l-size $m-size;
    text-align: center;
    width: 25rem;
}
.box-layout__title{
    margin: 0 0 $m-size 0;
    line-height: 1;
}