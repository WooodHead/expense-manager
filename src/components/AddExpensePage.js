import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import {add_expense } from '../actions/expenses';

export class AddExpensePage extends React.Component{
    onSubmit = (expense) => {
        this.props.addExpense(expense);
        this.props.history.push('/');
    };
    render() {
        return ( 
            <div> 
                <h2>Add Expense</h2>
                <ExpenseForm 
                    onSubmit={this.onSubmit}
                />
            </div> 
        );
    }
}

const mapDispatchToProps = (dispatch) => ({
    addExpense: (expense) => dispatch(add_expense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);