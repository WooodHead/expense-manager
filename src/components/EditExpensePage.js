import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { edit_expense, remove_expense } from '../actions/expenses';

const EditExpensePage = (props) => ( 
    <div> 
        <ExpenseForm 
            expense= {props.expense}
            onSubmit={(expense) => {
                //console.log(expense);
            props.dispatch(edit_expense(props.expense.id, expense));
            props.history.push('/');
            }}
        />
        <button onClick={(e) => {
            props.dispatch(remove_expense({id: props.expense.id }));
            props.history.push('/');
        }}>Remove Expense</button>

    </div> 
);

const mapStateToProps = (state,props) => {
    return {
        expense: state.expenses.find((expense) => expense.id === props.match.params.id )
    };
}

export default connect(mapStateToProps)(EditExpensePage);