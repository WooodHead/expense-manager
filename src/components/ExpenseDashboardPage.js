import React from 'react';
import ExpenseList from './ExpenseList';
import ExpeListFilter from './ExpenseListFilter';
import ExpensesSummary from './ExpensesSummary';

const ExpenseDashboardPage = () => ( 
    <div> 
        <ExpensesSummary />
        <ExpeListFilter />    
        <ExpenseList />
    </div> 
);

export default ExpenseDashboardPage;