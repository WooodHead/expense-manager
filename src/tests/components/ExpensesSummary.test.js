import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';
import testData from '../fixtures/expenses';

test('should render expenses summary for one expense', () => {
    const wrapper = shallow(<ExpensesSummary expensesCount={1} expensesTotal={234} />);
    expect(wrapper).toMatchSnapshot();
});

test('should render expenses summary for multiple expense', () => {
    const wrapper = shallow(<ExpensesSummary expensesCount={2} expensesTotal={23434234234}/>);
    expect(wrapper).toMatchSnapshot();
});