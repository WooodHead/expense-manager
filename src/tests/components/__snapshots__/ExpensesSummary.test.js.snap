// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expenses summary for multiple expense 1`] = `
<div>
  <h1>
     Viewing 
    2
     
    expenses
     totalling 
    $23,434,234,234.00
    .
  </h1>
</div>
`;

exports[`should render expenses summary for one expense 1`] = `
<div>
  <h1>
     Viewing 
    1
     
    expense
     totalling 
    $234.00
    .
  </h1>
</div>
`;
